src/JuMP.jl

```diff
@@ -719,22 +542,31 @@ end
 ##########################################################################
 # ConstraintRef
 # Reference to a constraint for retrieving solution info
-immutable ConstraintRef{M<:AbstractModel,T<:AbstractConstraint}
+# struct ConstraintRef{M<:AbstractModel,C <: MOICON}
+struct ConstraintRef{M<:AbstractModel,C} # TODO temporary
```

Comment:
Why is it temporary ?
I liked the `ConstraintRef{M<:AbstractModel, T<:AbstractConstraint}` design. This allows a constraint to be mapped to several MOICON. For example a SOS constraint is mapped to a VectorOfVariables/PSD constraint and a VectorAffineFunction/Zeros constraint.
With `ConstraintRef{M<:AbstractModel, T<:MOICON}`, we will be forced to define the Sum of squares cone as an `MOI.AbstractSet` and ask for every solver to support a bridge for it. The bridges cannot be implemented outside the solver (e.g. by defininte `addconstraint(m, f, s::SumOfSquaresCone)` since the solver needs to be aware the the initial constraint to implement attributes such as `ListOfConstraints`, `NumberOfConstraints`, `ConstraintFunction`, ...